sparse/_coo/numba_extension.py

```diff
@@ -0,0 +1,141 @@
+        self.coords_dtype = coords_dtype
+        self.ndim = ndim
+        super().__init__(
+            name="COOType[{!r}, {!r}, {!r}]".format(
+                numba.from_dtype(data_dtype), numba.from_dtype(coords_dtype), ndim
+            )
+        )
+
+    @property
+    def key(self):
+        return self.data_dtype, self.coords_dtype, self.ndim
+
+    @property
+    def data_type(self):
+        return numba.from_dtype(self.data_dtype)[:]
+
+    @property
+    def coords_type(self):
+        return numba.from_dtype(self.coords_dtype)[:, :]
+
+    @property
+    def shape_type(self):
+        return types.UniTuple(types.int64, self.ndim)
+
+
+@typeof_impl.register(COO)
+def _typeof_COO(val: COO, c) -> COOType:
+    return COOType(
+        data_dtype=val.data.dtype, coords_dtype=val.coords.dtype, ndim=val.ndim
+    )
+
+
+@register_model(COOType)
+class COOModel(models.StructModel):
+    def __init__(self, dmm, fe_type):
+        members = [
+            ("data", fe_type.data_type),
+            ("coords", fe_type.coords_type),
+            ("shape", fe_type.shape_type),
+        ]
+        models.StructModel.__init__(self, dmm, fe_type, members)
+
+
+@type_callable(COO)
+def type_COO(context):
+    def typer(coords, data, shape):
+        return COOType(
+            coords_dtype=numba.numpy_support.as_dtype(coords.dtype),
+            data_dtype=numba.numpy_support.as_dtype(data.dtype),
+            ndim=len(shape),
+        )
+
+    return typer
+
+
+@lower_builtin(COO, types.Any, types.Any, types.Any)
+def impl_COO(context, builder, sig, args):
+    typ = sig.return_type
+    coords, data, shape = args
+    coo = cgutils.create_struct_proxy(typ)(context, builder)
+    coo.coords = coords
+    coo.data = data
+    coo.shape = shape
+    return impl_ret_borrowed(context, builder, sig.return_type, coo._getvalue())
+
+
+@lower_constant(COOType)
+def lower_constant_COO(context, builder, typ, pyval):
+    coords = context.get_constant_generic(builder, typ.coords_type, pyval.coords)
+    data = context.get_constant_generic(builder, typ.data_type, pyval.data)
+    shape = context.get_constant_generic(builder, typ.shape_type, pyval.shape)
+    return impl_ret_borrowed(
+        context, builder, typ, cgutils.pack_struct(builder, (data, coords, shape))
+    )
+
+
+@unbox(COOType)
+def unbox_COO(typ: COOType, obj: COO, c) -> NativeValue:
+    data = c.pyapi.object_getattr_string(obj, "data")
+    coords = c.pyapi.object_getattr_string(obj, "coords")
+    shape = c.pyapi.object_getattr_string(obj, "shape")
+    coo = cgutils.create_struct_proxy(typ)(c.context, c.builder)
+    coo.coords = c.unbox(typ.coords_type, coords).value
+    coo.data = c.unbox(typ.data_type, data).value
+    coo.shape = c.unbox(typ.shape_type, shape).value
+    c.pyapi.decref(data)
+    c.pyapi.decref(coords)
+    c.pyapi.decref(shape)
+    is_error = cgutils.is_not_null(c.builder, c.pyapi.err_occurred())
+    return NativeValue(coo._getvalue(), is_error=is_error)
```

Comment:
@stuartarchibald: Do I need to call `is_not_null` after every single one of the python attribute accesses? What failure is this looking for here?

Reply:
The `is_error` is to signal something failed in unboxing and that there's an exception pending. I think you might be able to just use `is_error=c.pyapi.c_api_error()`, to make a check to the Python C-API for the error state as per `PyErr_Occurred`.

Reply:
I think I perhaps haven't conveyed my point. In c, this would look like:
```
coords = unbox(c_coords);
if (coords == NULL) {
    return NULL;
}
shape = unbox(c_shape);
if (shape == NULL) {
    Py_DECREF(coords);
    return NULL;
}
```
How would I translate that here?

Reply:
I think you'd need to replicate the first code block. Perhaps with things like (untested, I just typed this in):
```python
data = c.pyapi.object_getattr_string(obj, "data")
with cgutils.if_unlikely(c.builder,  cgutils.is_null(c.builder, data)):
  c.builder.ret(c.pyapi.get_null_object())
```
I don't recall from memory whether something also has to signal that an exception is pending. Perhaps try the above and see?

Reply:
That sounds roughly correct - if it works, then the interval example in yours docs should be updated to use it.

Reply:
With that example, what should I actually return from my python function body?

Reply:
> With that example, what should I actually return from my python function body?

Sorry, I'm not sure to what you are referring? If that branch is taken a NULL is returned?

Reply:
You can take a look at the LLVM IR that's being generated by breaking in this function and looking at `builder.module`.

Reply:
I mean what should the python function return, not what should the computation graph return

Reply:
This: `    return NativeValue(coo._getvalue(), is_error=is_error)` seems correct for what you have, whether there's extra error detection things you want to put into the `is_error` bit I cannot ascertain from what's present. Essentially, a `NativeValue` needs returning and in this case the first arg to needs to be a load-instruction where the thing being loaded points to the native representation of the thing being unboxed.

Reply:
Perhaps I can rephrase this another way: Provided every line that calls into Python's interpreter can potentially raise a Python exception, do we need to explicitly check for errors and perform the appropriate refcounting or will Numba skip the lines after the error and perform the correct refcounting for us?

Reply:
I think the conclusion was that we're on the hook for checking for python exceptions ourselves. My latest commits handle this, so marking this conversation as resolved.